Fix issues in RegisterForReflection#registerFullHierarchy() handling

gsmet opened the pull request:
We were passing to ReflectiveHierarchyBuildItem and index that solely contained the annotated class so it was doomed to fail and output a lot of warnings.

Fixes a few other minor infelicities in passing.

Fixes #38486

gsmet committed:
Fix issues in RegisterForReflection#registerFullHierarchy() handling

We were passing to ReflectiveHierarchyBuildItem and index that solely
contained the annotated class so it was doomed to fail and output a lot
of warnings.

Fixes a few other minor infelicities in passing.

Fixes #38486

gsmet committed:
Use ClassInfo#memberClasses() instead of using the ClassLoader

gsmet committed:
Reimplement @RegisterForReflection(registerFullHierarchy)

The idea is to leverage what we already have with the
ReflectiveHierarchyBuildItem infrastructure.

It simplifies things but I also did several changes for more
consistency:
- the @RegisterForReflection config is now passed to
  ReflectiveHierarchyBuildItem, which wasn't the case before
- we register the nested classes by default consistently with what
  @RegisterForReflection is doing by default

IMO, we need more consistency in this handling because we are using the
same annotation so the behavior should be consistent.
If you have one class registered for reflection and you decide to
register the hierarchy instead, the behavior should be the same.

gsmet commented:
I added a new commit that both simplifies the implementation and makes things a lot more consistent.

It changes a few behaviors though, one of each is that we are also now registering the nested classes for reflection when using `ReflectiveHierarchyBuildItem`.

@zakkak I'm interested in your feedback.

gsmet commented:
> Indeed, but I am not sure it's worth it. As commented in https://github.com/quarkusio/quarkus/issues/3883#issuecomment-2535991885:
>
> My main concern is that all this parameterization is tightly coupled with what's supported by GraalVM and this is a moving target (see https://github.com/oracle/graal/issues/9679).
> 
> So if we don't have an actual need to support them, perhaps we should avoid it.

I see where you are coming from... but I think we need consistency. We are using `@RegisterForReflection` here and it has these options so I don't think we can simply ignore them.

Also I have seen the issue with the more advanced options and I think we probably need to expose them too. I need to have a look at the plan they have but... it seems a bit counter productive to not expose fine grained things especially since the price of exposing all private things could be high in some cases.

> Although, this will probably impact the binary size, it seems like a reasonable default. On the other hand we haven't heard any complains related to its absence.

Yeah, I was on the verge but I think we need consistency, otherwise the behavior is very surprising given you would expect the `registerFullHierarchy` to add more things not less.

zakkak reviewed (approved):
LGTM. Any concerns I had were already discussed in the comments. I have no strong opinion on any of them, so I am OK with merging this.

Thank you @gsmet

zakkak commented:
> Also I have seen the issue with the more advanced options and I think we probably need to expose them too.

I think the design consensus was to try and stay as abstract as possible and avoid tightly coupling our API with what's offered by GraalVM, not sure if that has been revisited lately.

gsmet commented:
> It changes a few behaviors though, one of each is that we are also now registering the nested classes for reflection when using ReflectiveHierarchyBuildItem.

This a bad idea in the end. I need to fix that.

gsmet committed:
Ignore nested classes by default for ReflectiveHierarchyBuildItem

This is consistent with the original behavior.

For @RegisterForReflection, we honor the default value of ignoreNested()
but I added a comment pointing to it in the Javadoc.

gsmet commented:
I added a commit to get back to the ignoring nested classes by default when using `ReflectiveHierarchyBuildItem`.

We honor `ignoreNested` for `@RegisterForReflection` now though, I added a comment in the Javadoc to make it extra clear.
